File: sdk/go/fit/fit.go
// Package fit 在 request/response Snapshot 之间提供可选的 bid↔request 契合检查。
//
// 非 JSON Schema，也非 LightGate。返回软性的 FitResult 发现项。
// OK() 仅表示无 ERROR；WARN（低价/屏蔽/超时等）仍可能 OK；政策拒投请读 Warnings()/Has。
package fit

import (
	"strconv"
	"strings"

	"github.com/oakrtb/openrtb/sdk/go/view"
	openrtb "github.com/oakrtb/openrtb/sdk/go/oakrtb/v2"
)

// ImpReadyMtype 检查所选 markup 类型是否存在于 Imp 且具备构建器级规格。
func ImpReadyMtype(imp *view.ImpView, mtype int32) FitResult {
	if imp == nil {
		return FitResult{Issues: []FitIssue{{
			Code: CodeMalformed, Severity: SeverityError, Path: "imp", Message: "nil ImpView",
		}}}
	}
	path := "imp[" + imp.ID + "]"
	if mtype < 1 || mtype > 4 {
		return FitResult{Issues: []FitIssue{{
			Code: CodeMtypeUnknown, Severity: SeverityError, Path: path + ".mtype",
			Message: "mtype must be 1–4 for ImpReady",
		}}}
	}
	return impReadyChosen(imp, view.MarkupFromMtype(openrtb.MarkupType(mtype)), path)
}

// ImpReadyMarkup 与 ImpReadyMtype 类似，但使用单比特 MarkupMask 指定类型。
func ImpReadyMarkup(imp *view.ImpView, chosen view.MarkupMask) FitResult {
	if imp == nil {
		return FitResult{Issues: []FitIssue{{
			Code: CodeMalformed, Severity: SeverityError, Path: "imp", Message: "nil ImpView",
		}}}
	}
	path := "imp[" + imp.ID + "]"
	if chosen.Count() != 1 {
		return FitResult{Issues: []FitIssue{{
			Code: CodeMultiNeedsChoice, Severity: SeverityError, Path: path,
			Message: "chosen MarkupMask must be exactly one bit",
		}}}
	}
	return impReadyChosen(imp, chosen, path)
}

func impReadyChosen(imp *view.ImpView, chosen view.MarkupMask, path string) FitResult {
	var issues []FitIssue
	if !imp.Markup.Has(chosen) {
		return FitResult{Issues: []FitIssue{{
			Code: CodeFormatNotOnImp, Severity: SeverityError, Path: path,
			Message: "chosen format is not present on Imp",
		}}}
	}
	if chosen.HasBanner() {
		b := imp.Banner
		sizeOK := b != nil && ((b.W > 0 && b.H > 0) || len(b.Format) > 0)
		if !sizeOK {
			issues = append(issues, FitIssue{
				Code: CodeBannerSizeMissing, Severity: SeverityError, Path: path + ".banner",
				Message: "banner needs w/h or format[]",
			})
		}
	}
	if chosen.HasVideo() {
		v := imp.Video
		if v == nil || len(v.Mimes) == 0 {
			issues = append(issues, FitIssue{
				Code: CodeVideoMimesMissing, Severity: SeverityError, Path: path + ".video.mimes",
				Message: "video.mimes is required",
			})
		} else {
			if v.Minduration == 0 && v.Maxduration == 0 {
				issues = append(issues, FitIssue{
					Code: CodeVideoDurationUnset, Severity: SeverityWarn, Path: path + ".video",
					Message: "video minduration/maxduration unset",
				})
			}
			if len(v.Protocols) == 0 {
				issues = append(issues, FitIssue{
					Code: CodeVideoProtocolsUnset, Severity: SeverityWarn, Path: path + ".video.protocols",
					Message: "video.protocols unset",
				})
			}
		}
	}
	if chosen.HasAudio() {
		a := imp.Audio
		if a == nil || len(a.Mimes) == 0 {
			issues = append(issues, FitIssue{
				Code: CodeAudioMimesMissing, Severity: SeverityError, Path: path + ".audio.mimes",
				Message: "audio.mimes is required",
			})
		}
	}
	if chosen.HasNative() {
		n := imp.Native
		if n == nil || strings.TrimSpace(n.Request) == "" {
			issues = append(issues, FitIssue{
				Code: CodeNativeRequestMissing, Severity: SeverityError, Path: path + ".native.request",
				Message: "native.request is required",
			})
		}
	}
	return FitResult{Issues: issues}
}

// Bid 将单个 Bid 与请求快照做 Fit。无响应货币时跳过底价比较；完整检查请用 Response。
func Bid(req *view.RequestSnapshot, bid *openrtb.Bid) FitResult {
	return bidAt(req, bid, "", "bid")
}

// Response 对整个 BidResponse 做匹配检查；空 seatbid 视为通过（可附加 PAST_DEADLINE 警告）。
func Response(req *view.RequestSnapshot, res *openrtb.BidResponse) FitResult {
	var issues []FitIssue
	if req == nil || res == nil {
		return FitResult{Issues: []FitIssue{{
			Code: CodeMalformed, Severity: SeverityError, Path: "", Message: "nil req or res",
		}}}
	}
	if req.PastDeadline() {
		issues = append(issues, FitIssue{
			Code: CodePastDeadline, Severity: SeverityWarn, Path: "tmax",
			Message: "past request deadline (85% of tmax)",
		})
	}
	if len(res.Seatbid) == 0 {
		return FitResult{Issues: issues}
	}
	// Same blank rule as LightGate / checkFloor: trim; blank skips CUR + floor.
	resCur := strings.TrimSpace(res.Cur)
	if resCur != "" {
		allowed := false
		for _, c := range req.Currencies() {
			if strings.EqualFold(resCur, strings.TrimSpace(c)) {
				allowed = true
				break
			}
		}
		if !allowed {
			issues = append(issues, FitIssue{
				Code: CodeCurNotAllowed, Severity: SeverityWarn, Path: "BidResponse.cur",
				Message: "response currency not in BidRequest.cur",
			})
		}
	}
	for i, sb := range res.Seatbid {
		if sb == nil {
			issues = append(issues, FitIssue{
				Code: CodeMalformed, Severity: SeverityError, Path: "seatbid[" + itoa(i) + "]",
				Message: "seatbid entry is nil",
			})
			continue
		}
		if len(sb.Bid) == 0 {
			issues = append(issues, FitIssue{
				Code: CodeMalformed, Severity: SeverityError, Path: "seatbid[" + itoa(i) + "].bid",
				Message: "seatbid.bid must be a non-empty array",
			})
			continue
		}
		for j, bid := range sb.Bid {
			if bid == nil {
				issues = append(issues, FitIssue{
					Code: CodeMalformed, Severity: SeverityError,
					Path: "seatbid[" + itoa(i) + "].bid[" + itoa(j) + "]",
					Message: "bid is nil",
				})
				continue
			}
			path := "seatbid[" + itoa(i) + "].bid[" + itoa(j) + "]"
			one := bidAt(req, bid, resCur, path)
			issues = append(issues, one.Issues...)
		}
	}
	return FitResult{Issues: issues}
}

func bidAt(req *view.RequestSnapshot, bid *openrtb.Bid, responseCur, path string) FitResult {
	var issues []FitIssue
	if req == nil || bid == nil {
		return FitResult{Issues: []FitIssue{{
			Code: CodeMalformed, Severity: SeverityError, Path: path, Message: "nil req or bid",
		}}}
	}
	imp := req.FindImp(bid.Impid)
	if imp == nil {
		return FitResult{Issues: []FitIssue{{
			Code: CodeImpNotFound, Severity: SeverityError, Path: path + ".impid",
			Message: "impid not found in request",
		}}}
	}
	mtype := int32(bid.GetMtype())
	markup := imp.Markup

	if mtype >= 500 {
		issues = append(issues, FitIssue{
			Code: CodeMtypeVendor, Severity: SeverityWarn, Path: path + ".mtype",
			Message: "vendor mtype >=500",
		})
	} else if mtype != 0 && (mtype < 1 || mtype > 4) {
		issues = append(issues, FitIssue{
			Code: CodeMtypeUnknown, Severity: SeverityError, Path: path + ".mtype",
			Message: "mtype must be 0–4 or >=500",
		})
	} else if mtype == 0 && markup.Count() > 1 {
		issues = append(issues, FitIssue{
			Code: CodeMtypeRequired, Severity: SeverityError, Path: path + ".mtype",
			Message: "multi-format Imp requires Bid.mtype",
		})
	} else if mtype >= 1 && mtype <= 4 {
		chosen := view.MarkupFromMtype(openrtb.MarkupType(mtype))
		if !markup.Has(chosen) {
			issues = append(issues, FitIssue{
				Code: CodeMtypeMismatch, Severity: SeverityError, Path: path + ".mtype",
				Message: "mtype does not match Imp formats",
			})
		}
	}

	eff := view.MarkupNone
	if mtype >= 1 && mtype <= 4 {
		eff = view.MarkupFromMtype(openrtb.MarkupType(mtype))
	} else if markup.Count() == 1 {
		eff = markup
	}

	issues = append(issues, checkFloor(imp, bid, responseCur, path)...)
	issues = append(issues, checkAttr(imp, bid, eff, path)...)
	issues = append(issues, checkBlocks(req, bid, path)...)
	return FitResult{Issues: issues}
}

func checkFloor(imp *view.ImpView, bid *openrtb.Bid, responseCur, path string) []FitIssue {
	floor := imp.BidFloor
	if floor <= 0 {
		return nil
	}
	// Require both response cur and bidfloorcur before numeric compare (Fit.Bid has no response cur).
	respCur := strings.TrimSpace(responseCur)
	floorCur := strings.TrimSpace(imp.BidFloorCur)
	if respCur == "" || floorCur == "" {
		return nil
	}
	if !strings.EqualFold(respCur, floorCur) {
		return []FitIssue{{
			Code: CodeFloorCurDiff, Severity: SeverityWarn, Path: path + ".price",
			Message: "bid currency differs from imp.bidfloorcur; skip floor compare",
		}}
	}
	if bid.Price < floor {
		return []FitIssue{{
			Code: CodePriceBelowFloor, Severity: SeverityWarn, Path: path + ".price",
			Message: "price below imp.bidfloor",
		}}
	}
	return nil
}

func checkAttr(imp *view.ImpView, bid *openrtb.Bid, eff view.MarkupMask, path string) []FitIssue {
	if len(bid.Attr) == 0 || eff == view.MarkupNone {
		return nil
	}
	battr := battrFor(imp, eff)
	if len(battr) == 0 {
		return nil
	}
	blocked := map[int32]struct{}{}
	for _, v := range battr {
		blocked[v] = struct{}{}
	}
	for _, a := range bid.Attr {
		if _, ok := blocked[a]; ok {
			return []FitIssue{{
				Code: CodeAttrBlocked, Severity: SeverityWarn, Path: path + ".attr",
				Message: "bid.attr intersects format battr",
			}}
		}
	}
	return nil
}

func battrFor(imp *view.ImpView, eff view.MarkupMask) []int32 {
	if eff.HasBanner() && imp.Banner != nil {
		return imp.Banner.Battr
	}
	if eff.HasVideo() && imp.Video != nil {
		return imp.Video.Battr
	}
	if eff.HasAudio() && imp.Audio != nil {
		return imp.Audio.Battr
	}
	if eff.HasNative() && imp.Native != nil {
		return imp.Native.Battr
	}
	return nil
}

func checkBlocks(req *view.RequestSnapshot, bid *openrtb.Bid, path string) []FitIssue {
	var issues []FitIssue
	if len(req.Shared.Badv) > 0 {
		block := lowerSet(req.Shared.Badv)
		for _, d := range bid.Adomain {
			if _, ok := block[strings.ToLower(d)]; ok {
				issues = append(issues, FitIssue{
					Code: CodeAdomainBlocked, Severity: SeverityWarn, Path: path + ".adomain",
					Message: "adomain hit BidRequest.badv",
				})
				break
			}
		}
	}
	if len(req.Shared.Bapp) > 0 && strings.TrimSpace(bid.Bundle) != "" {
		block := lowerSet(req.Shared.Bapp)
		if _, ok := block[strings.ToLower(strings.TrimSpace(bid.Bundle))]; ok {
			issues = append(issues, FitIssue{
				Code: CodeBundleBlocked, Severity: SeverityWarn, Path: path + ".bundle",
				Message: "bundle hit BidRequest.bapp",
			})
		}
	}
	if len(req.Shared.Bcat) > 0 {
		block := lowerSet(req.Shared.Bcat)
		for _, c := range bid.Cat {
			if _, ok := block[strings.ToLower(c)]; ok {
				issues = append(issues, FitIssue{
					Code: CodeCatBlocked, Severity: SeverityWarn, Path: path + ".cat",
					Message: "cat hit BidRequest.bcat",
				})
				break
			}
		}
	}
	return issues
}

func lowerSet(in []string) map[string]struct{} {
	out := make(map[string]struct{}, len(in))
	for _, s := range in {
		out[strings.ToLower(s)] = struct{}{}
	}
	return out
}

func itoa(i int) string {
	return strconv.Itoa(i)
}

File: sdk/java/src/main/java/com/oakrtb/sdk/fit/Fit.java
package com.oakrtb.sdk.fit;

import com.oakrtb.openrtb.v2.Audio;
import com.oakrtb.openrtb.v2.Banner;
import com.oakrtb.openrtb.v2.Bid;
import com.oakrtb.openrtb.v2.BidResponse;
import com.oakrtb.openrtb.v2.Native;
import com.oakrtb.openrtb.v2.SeatBid;
import com.oakrtb.openrtb.v2.Video;
import com.oakrtb.sdk.inspect.MarkupMask;
import com.oakrtb.sdk.inspect.RequestInspect;
import com.oakrtb.sdk.inspect.RequestPipeline;
import com.oakrtb.sdk.inspect.ResponseInspect;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;

/**
 * RequestInspect 与 ResponseInspect 之间的可选 bid↔request 一致性检查。
 *
 * <p>非 JSON Schema，也非 LightGate；返回软性 {@link FitResult}（不抛异常）。
 * 调用方自行决定是否因 ERROR 拒绝或处理 WARN。
 */
public final class Fit {
  private Fit() {}

  /**
   * 检查所选 markup 类型是否在 Imp 上存在，且具备 OakRTB 构建器级 spec。
   *
   * @param imp Imp 视图
   * @param mtype OpenRTB mtype（1–4）
   * @return Fit 结果
   */
  public static FitResult impReady(RequestInspect.ImpView imp, int mtype) {
    Objects.requireNonNull(imp, "imp");
    List<FitIssue> issues = new ArrayList<>();
    String path = "imp[" + imp.id() + "]";
    MarkupMask chosen = ResponseInspect.markupFromMtypeValue(mtype);
    if (mtype < 1 || mtype > 4) {
      issues.add(
          new FitIssue(
              IssueCode.MTYPE_UNKNOWN,
              Severity.ERROR,
              path + ".mtype",
              "mtype must be 1–4 for impReady"));
      return FitResult.of(issues);
    }
    return impReadyChosen(imp, chosen, path, issues);
  }

  /**
   * 与 {@link #impReady(RequestInspect.ImpView, int)} 相同，但使用单 bit {@link MarkupMask}。
   *
   * @param imp Imp 视图
   * @param chosen 须恰好一种格式的掩码
   * @return Fit 结果
   */
  public static FitResult impReady(RequestInspect.ImpView imp, MarkupMask chosen) {
    Objects.requireNonNull(imp, "imp");
    Objects.requireNonNull(chosen, "chosen");
    List<FitIssue> issues = new ArrayList<>();
    String path = "imp[" + imp.id() + "]";
    if (chosen.count() != 1) {
      issues.add(
          new FitIssue(
              IssueCode.MULTI_NEEDS_CHOICE,
              Severity.ERROR,
              path,
              "chosen MarkupMask must be exactly one bit"));
      return FitResult.of(issues);
    }
    return impReadyChosen(imp, chosen, path, issues);
  }

  private static FitResult impReadyChosen(
      RequestInspect.ImpView imp, MarkupMask chosen, String path, List<FitIssue> issues) {
    if (!imp.markup().has(chosen.bits())) {
      issues.add(
          new FitIssue(
              IssueCode.FORMAT_NOT_ON_IMP,
              Severity.ERROR,
              path,
              "chosen format is not present on Imp"));
      return FitResult.of(issues);
    }
    if (chosen.hasBanner()) {
      Banner b = imp.banner();
      boolean sizeOk =
          b != null && ((b.getW() > 0 && b.getH() > 0) || b.getFormatCount() > 0);
      if (!sizeOk) {
        issues.add(
            new FitIssue(
                IssueCode.BANNER_SIZE_MISSING,
                Severity.ERROR,
                path + ".banner",
                "banner needs w/h or format[]"));
      }
    }
    if (chosen.hasVideo()) {
      Video v = imp.video();
      if (v == null || v.getMimesCount() == 0) {
        issues.add(
            new FitIssue(
                IssueCode.VIDEO_MIMES_MISSING,
                Severity.ERROR,
                path + ".video.mimes",
                "video.mimes is required"));
      } else {
        if (v.getMinduration() == 0 && v.getMaxduration() == 0) {
          issues.add(
              new FitIssue(
                  IssueCode.VIDEO_DURATION_UNSET,
                  Severity.WARN,
                  path + ".video",
                  "video minduration/maxduration unset"));
        }
        if (v.getProtocolsCount() == 0) {
          issues.add(
              new FitIssue(
                  IssueCode.VIDEO_PROTOCOLS_UNSET,
                  Severity.WARN,
                  path + ".video.protocols",
                  "video.protocols unset"));
        }
      }
    }
    if (chosen.hasAudio()) {
      Audio a = imp.audio();
      if (a == null || a.getMimesCount() == 0) {
        issues.add(
            new FitIssue(
                IssueCode.AUDIO_MIMES_MISSING,
                Severity.ERROR,
                path + ".audio.mimes",
                "audio.mimes is required"));
      }
    }
    if (chosen.hasNative()) {
      Native n = imp.nativeAd();
      if (n == null || n.getRequest().isBlank()) {
        issues.add(
            new FitIssue(
                IssueCode.NATIVE_REQUEST_MISSING,
                Severity.ERROR,
                path + ".native.request",
                "native.request is required"));
      }
    }
    return FitResult.of(issues);
  }

  /**
   * 将单条 Bid 与请求快照做 Fit（不含响应级货币检查）。
   *
   * @param req 请求管道快照
   * @param bid 出价
   * @return Fit 结果
   */
  public static FitResult bidFit(RequestPipeline.Snapshot req, Bid bid) {
    return bidFit(req, bid, null, "bid");
  }

  /**
   * 对整个 BidResponse 做 Fit。空 seatbid（wire no-bid）返回空 ok 结果；
   * 必要时在响应级追加 PAST_DEADLINE / CUR_NOT_ALLOWED。
   *
   * @param req 请求管道快照
   * @param res 出价响应
   * @return Fit 结果
   */
  public static FitResult responseFit(RequestPipeline.Snapshot req, BidResponse res) {
    Objects.requireNonNull(req, "req");
    Objects.requireNonNull(res, "res");
    List<FitIssue> issues = new ArrayList<>();
    if (req.pastDeadline()) {
      issues.add(
          new FitIssue(
              IssueCode.PAST_DEADLINE,
              Severity.WARN,
              "tmax",
              "past request deadline (85% of tmax)"));
    }
    if (res.getSeatbidCount() == 0) {
      return FitResult.of(issues);
    }
    String resCur = res.getCur();
    if (resCur != null && !resCur.isBlank()) {
      boolean allowed = false;
      for (String c : req.currencies()) {
        if (resCur.equalsIgnoreCase(c)) {
          allowed = true;
          break;
        }
      }
      if (!allowed) {
        issues.add(
            new FitIssue(
                IssueCode.CUR_NOT_ALLOWED,
                Severity.WARN,
                "BidResponse.cur",
                "response currency not in BidRequest.cur"));
      }
    }
    for (int i = 0; i < res.getSeatbidCount(); i++) {
      SeatBid sb = res.getSeatbid(i);
      for (int j = 0; j < sb.getBidCount(); j++) {
        String path = "seatbid[" + i + "].bid[" + j + "]";
        FitResult one = bidFit(req, sb.getBid(j), resCur, path);
        issues.addAll(one.issues());
      }
    }
    return FitResult.of(issues);
  }

  private static FitResult bidFit(
      RequestPipeline.Snapshot req, Bid bid, String responseCur, String path) {
    Objects.requireNonNull(req, "req");
    Objects.requireNonNull(bid, "bid");
    List<FitIssue> issues = new ArrayList<>();
    Optional<RequestInspect.ImpView> found = req.findImp(bid.getImpid());
    if (found.isEmpty()) {
      issues.add(
          new FitIssue(
              IssueCode.IMP_NOT_FOUND,
              Severity.ERROR,
              path + ".impid",
              "impid not found in request"));
      return FitResult.of(issues);
    }
    RequestInspect.ImpView imp = found.get();
    int mtype = bid.getMtypeValue();
    MarkupMask markup = imp.markup();

    if (mtype >= 500) {
      issues.add(
          new FitIssue(
              IssueCode.MTYPE_VENDOR,
              Severity.WARN,
              path + ".mtype",
              "vendor mtype >=500"));
    } else if (mtype != 0 && (mtype < 1 || mtype > 4)) {
      issues.add(
          new FitIssue(
              IssueCode.MTYPE_UNKNOWN,
              Severity.ERROR,
              path + ".mtype",
              "mtype must be 0–4 or >=500"));
    } else if (mtype == 0 && markup.count() > 1) {
      issues.add(
          new FitIssue(
              IssueCode.MTYPE_REQUIRED,
              Severity.ERROR,
              path + ".mtype",
              "multi-format Imp requires Bid.mtype"));
    } else if (mtype >= 1 && mtype <= 4) {
      MarkupMask chosen = ResponseInspect.markupFromMtypeValue(mtype);
      if (!markup.has(chosen.bits())) {
        issues.add(
            new FitIssue(
                IssueCode.MTYPE_MISMATCH,
                Severity.ERROR,
                path + ".mtype",
                "mtype does not match Imp formats"));
      }
    }

    // Effective format for blocklist checks
    MarkupMask eff =
        (mtype >= 1 && mtype <= 4)
            ? ResponseInspect.markupFromMtypeValue(mtype)
            : (markup.count() == 1 ? markup : MarkupMask.of(MarkupMask.NONE));

    checkFloor(issues, imp, bid, responseCur, path);
    checkAttr(issues, imp, bid, eff, path);
    checkBlocks(issues, req, bid, path);
    return FitResult.of(issues);
  }

  private static void checkFloor(
      List<FitIssue> issues,
      RequestInspect.ImpView imp,
      Bid bid,
      String responseCur,
      String path) {
    double floor = imp.bidFloor();
    if (floor <= 0) {
      return;
    }
    String floorCur = imp.bidFloorCur();
    if (responseCur != null
        && !responseCur.isBlank()
        && floorCur != null
        && !floorCur.isBlank()
        && !responseCur.equalsIgnoreCase(floorCur)) {
      issues.add(
          new FitIssue(
              IssueCode.FLOOR_CUR_DIFF,
              Severity.WARN,
              path + ".price",
              "bid currency differs from imp.bidfloorcur; skip floor compare"));
      return;
    }
    if (bid.getPrice() < floor) {
      issues.add(
          new FitIssue(
              IssueCode.PRICE_BELOW_FLOOR,
              Severity.WARN,
              path + ".price",
              "price below imp.bidfloor"));
    }
  }

  private static void checkAttr(
      List<FitIssue> issues,
      RequestInspect.ImpView imp,
      Bid bid,
      MarkupMask eff,
      String path) {
    if (bid.getAttrCount() == 0 || eff.bits() == MarkupMask.NONE) {
      return;
    }
    List<Integer> battr = battrFor(imp, eff);
    if (battr.isEmpty()) {
      return;
    }
    Set<Integer> blocked = new HashSet<>(battr);
    for (int a : bid.getAttrList()) {
      if (blocked.contains(a)) {
        issues.add(
            new FitIssue(
                IssueCode.ATTR_BLOCKED,
                Severity.WARN,
                path + ".attr",
                "bid.attr intersects format battr"));
        return;
      }
    }
  }

  private static List<Integer> battrFor(RequestInspect.ImpView imp, MarkupMask eff) {
    if (eff.hasBanner() && imp.banner() != null) {
      return imp.banner().getBattrList();
    }
    if (eff.hasVideo() && imp.video() != null) {
      return imp.video().getBattrList();
    }
    if (eff.hasAudio() && imp.audio() != null) {
      return imp.audio().getBattrList();
    }
    if (eff.hasNative() && imp.nativeAd() != null) {
      return imp.nativeAd().getBattrList();
    }
    return List.of();
  }

  private static void checkBlocks(
      List<FitIssue> issues, RequestPipeline.Snapshot req, Bid bid, String path) {
    List<String> badv = req.shared().badv();
    if (!badv.isEmpty()) {
      Set<String> block = toLowerSet(badv);
      for (String d : bid.getAdomainList()) {
        if (d != null && block.contains(d.toLowerCase(Locale.ROOT))) {
          issues.add(
              new FitIssue(
                  IssueCode.ADOMAIN_BLOCKED,
                  Severity.WARN,
                  path + ".adomain",
                  "adomain hit BidRequest.badv"));
          break;
        }
      }
    }
    List<String> bapp = req.shared().bapp();
    if (!bapp.isEmpty() && bid.getBundle() != null && !bid.getBundle().isBlank()) {
      Set<String> block = toLowerSet(bapp);
      if (block.contains(bid.getBundle().toLowerCase(Locale.ROOT))) {
        issues.add(
            new FitIssue(
                IssueCode.BUNDLE_BLOCKED,
                Severity.WARN,
                path + ".bundle",
                "bundle hit BidRequest.bapp"));
      }
    }
    List<String> bcat = req.shared().bcat();
    if (!bcat.isEmpty()) {
      Set<String> block = toLowerSet(bcat);
      for (String c : bid.getCatList()) {
        if (c != null && block.contains(c.toLowerCase(Locale.ROOT))) {
          issues.add(
              new FitIssue(
                  IssueCode.CAT_BLOCKED,
                  Severity.WARN,
                  path + ".cat",
                  "cat hit BidRequest.bcat"));
          break;
        }
      }
    }
  }

  private static Set<String> toLowerSet(List<String> in) {
    Set<String> out = new HashSet<>();
    for (String s : in) {
      if (s != null) {
        out.add(s.toLowerCase(Locale.ROOT));
      }
    }
    return out;
  }
}
